### ISSUE TITLE
Feature: Accessing request and response objects within component class

### ISSUE THREAD

**Issue by WhoAteDaCake** [analogjs/analog#586]
### Which scope/s are relevant/related to the feature request?

router

### Information

Adding an ability to access request and response objects, would allow for server side authentication. As an example next js provides an exported function:

```ts
export const getServerSideProps = withSession(async function ({ req, res }) {
  const { user } = req.session
 
  if (!user) {
    return {
      redirect: {
        destination: '/login',
        permanent: false,
      },
    }
  }
 
  return {
    props: { user },
  }
})
 ```

It follows a similar pattern to [component.server.ts](https://github.com/analogjs/analog/blob/main/apps/analog-app/src/app/pages/(home).server.ts) files. The signature of the `load` function could be updated to return redirects, or another function to be exposed in order to not break the compatability.

### Describe any alternatives/workarounds you're currently using

No alternative right now.

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

**Comment by brandonroberts** [comment]
This seems reasonable to me. What would the change look like for the `load` function vs a new function?

We could also expose these as injectable tokens or functions.

**Comment by WhoAteDaCake** [comment]
@brandonroberts 
If the function is changed, it would require a more concrete return type. It could be very similar to next.js:

```
export type Redirect =
  | {
      statusCode: 301 | 302 | 303 | 307 | 308
      destination: string
      basePath?: false
    }
  | {
      permanent: boolean
      destination: string
      basePath?: false
    }

export type LoadResult<T> =
  | { data: T }
  | { redirect: Redirect }
  | { notFound: true }

export type LoadFuc = (pageServerLoad: PageServerLoad) => Promise<LoadResult<T>>
```

If a new function is added it could be as simple as a middleware handler:

```
// Alternative names: onRequest, request
export function beforeLoad(pageServerLoad: PageServerLoad, next: () => void) {
}
``

> We could also expose these as injectable tokens or functions.
From my opinion, I'd prefer to have the routing/server logic with a bit of  separation from UI component, to avoid having to deal with `if (isBrowser)` like logic

**Comment by brandonroberts** [comment]
That seems reasonable to me. We could keep `T` instead of `{ data:  T }` but either works.

**Comment by WhoAteDaCake** [comment]
So would your preference be to update the existing load function?

The only reason I suggested using `{ data:  T }`, was to have a concrete type, which doesn't clash with custom handling logic. Someone might return `{notFound: true}` to show that user is not found and then get confused by 404 being returned.

**Comment by brandonroberts** [comment]
Yes, that's what my preference would be. As long as we can narrow the type of the client side usage where the injectLoad function will infer the data type.

**Comment by brandonroberts** [comment]
Support for Request, Response is added in https://github.com/analogjs/analog/pull/806 and included in the `0.2.29` release
